### DISCUSSION TITLE
幽霊役職：アサイン周りの仕様

### DISCUSSION THREAD

**Discussion by yukieiji** [yukieiji/ExtremeRoles#46] [Ideas]
# 幽霊役職のアサイン周りの仕様議論場所
大元のチケット：#44

- 新仕様
  - 陣営をまたいだアサインはされない
  - 幽霊役職のアサインされた人は会議中に天使のアイコンが付く(オプションで切り替え化)
  - アサインされる幽霊役職の元役職の指定があったりする
    - 元役職の判定は一番上の役職を見る(例：ラバーズ+サイドキックならラバーズだけを見る)
  - 幽霊役職と通常役職のコンビ役職も存在する
  - 実装コストが高すぎるため、死因によるアサインフィルターは今後一切実装しない
    - もし実現できたとしても「追放」と「キル」以外の状態を取ることはキツイ

- 2022/05/17T23:58以前ディスカッションに記載されていた旧仕様
  - チケットに上がった段階ので仕様
    - アサインされた人は天使のアイコンをつける
    - 陣営をまたいだアサインはされない
    - 基本的にバニラの「守護天使」と同じアサインロジックを使用

**Comment by hijiki55** [comment]
アサイン条件にどの程度柔軟性を持たせられるかが知りたいです
 - 死亡の原因をアサイン条件に加えることは可能ですか？(キルで死亡した場合のみアサインされるなど)
 - 個々の役職をアサイン条件に加えることは可能ですか？(素クルーや素インポスターからのみアサインされるなど)

インポ第三の食い合い展開のクッションなどに使えそうに思いました

**Reply by yukieiji** [reply]
> アサイン条件にどの程度柔軟性を持たせられるかが知りたいです

幽霊役職のプログラムは1文字も書いていないので制限は無く、柔軟性は無限大です(例に上げてる物は基本的に出来ます)

今は仕様を考えている段階なので、制限は無いものだと考えて仕様を考えていただいて構いません
私の考えている幽霊役職の仕様はバニラのアサインロジックをそのまま持ってくる予定です

**Reply by yukieiji** [reply]
一応確認なんですけど、この仕様がほしいという解釈でよろしいでしょうか

**Reply by hijiki55** [reply]
> 一応確認なんですけど、この仕様がほしいという解釈でよろしいでしょうか

はい、そうです
ただ、守護天使のような早い者勝ちでアサイン枠を取り合う形のところに、より狭いアサイン条件を加える場合、アサインの優先順位仕様などを考える必要が出てくるという認識はこちらとしても持っています。その辺はお任せしたいです

**Reply by yukieiji** [reply]
> ただ、守護天使のような早い者勝ちでアサイン枠を取り合う形のところに、より狭いアサイン条件を加える場合、アサインの優先順位仕様などを考える必要が出てくるという認識はこちらとしても持っています。その辺はお任せしたいです

そこら辺は内部仕様に近いので今は考えなくていい段階です

**Reply by yukieiji** [reply]
実装コスト的に死因を分けることは難しそうですが問題ないでしょうか

**Comment by yukieiji** [comment]
実装に先駆けて仕様書的なものを作るため
5月16日午前0時までに追加意見等がない場合は一旦この仕様での実装可能調査や詳しい内部仕様を決定します(決定後新しいディスカッションを作ります)

**Comment by UzomuZo** [comment]
幽霊役職のアサインがどういう風になっているかわからないのですが、
たとえばクルー役職最大4人で、クルー幽霊役職2，生身(？)のクルー役職6で全部50%にしていたら幽霊2入ることもあるのでしょうか？
同陣営に2職アサインされないのであれば問題ありませんが、もしされるならば同時に条件を満たした場合はどちらの幽霊役職になるのでしょうか？

あとは幽霊役職が増えていったときのことを考えると、生身のクルー役職は6つから4つランダム、幽霊は3つから1つランダムみたいな設定ができるように別で人数設定できた方がいい気がします。

**Reply by yukieiji** [reply]
基本的に幽霊役職は別の設定で一括管理する予定です。「クルー幽霊役職数」とか新規に設定を実装します。なので同じ設定が混在するってことはありません

**Reply by UzomuZo** [reply]
わかりました、ありがとうございます。

**Comment by Inumiya** [comment]
生前役職Xと幽霊役職Yをコンビ役職にして、Xがアサインされた場合、必ずXは死後Yになる仕様があるといいなと思いました
生前役職の能力で得た情報を元に死後役職の能力を使ったり、生前役職は弱いけれどその分死後役職が強いなど役職を考えるときに幅が広がるかと思います

**Comment by yukieiji** [comment]
幽霊役職の仕様書を記載するため一時的にロックします

**Reply by yukieiji** [reply]
一時的な外部仕様が完成したためディスカッションを再度開放します

**Comment by melonjagaimo** [comment]
質問が2つあります
１：この役職は死亡しても幽霊役職に変化しない、というのは可能でしょうか
例えばマーリンが死亡して幽霊役職に変化した場合、アサシンはマーリン的中による勝利が不可能になってしまいます

２：特定の役職からのみ派生する幽霊役職は可能でしょうか
例えばラバーズが「片方が死亡しても後追いせず、キルボタンを得てニュートラル陣営になる」設定の場合、先に死亡した方が専用の幽霊役職になって残された相方の支援ができるとか

**Reply by yukieiji** [reply]
> この役職は死亡しても幽霊役職に変化しない、というのは可能でしょうか
例えばマーリンが死亡して幽霊役職に変化した場合、アサシンはマーリン的中による勝利が不可能になってしまいます

まず質問の解答です。それは基本的には考えてないです
例に書いてある部分を読む限り、幽霊役職は既存役職を置き換えるみたいな感じに見えたのだと思いますが、幽霊役職は既存の役職に追加してアサインされます。例：ラバーズ+ウォッチドッグ => 死亡後：守護天使(ラバーズ+ウォッチドッグ)みたいな扱いになります

> 特定の役職からのみ派生する幽霊役職は可能でしょうか

これは新仕様のところの「アサインされる元役職の指定があったりする」があるので可能ですし考慮した実装になります
